### ISSUE TITLE
Loop starvation time

### ISSUE THREAD

**Issue by logidelic** [ibmruntimes/yieldable-json#25]
Is the `5ms` mentioned in the documentation somehow configurable internally? If so, it would be nice to expose this as an optional configuration parameter... As you mention, the "definition of concurrency is subjective" and some systems actually require an even smaller threshold...

**Comment by gireeshpunathil** [comment]
@logidelic - thanks for reviewing that, and the issue. `5 ms` is empirically drawn value, neither fixed nor tunable.

However, we take an input  called `intensity` for the APIs. This essentially acts as a counter, while traversing through deeply nested structures, when the count of traversal reaches the `intensity` value, we return back from the JSON API (with a yield clause, so that we can resume from there later), and attend to the rest of the work for a while.

The value of `5 ms` is obtained from a normal linux system when we used `512` as the intensity. So we used this default value.

As you say, if the system, the nature of the JSON, the responsiveness SLA etc. changes and you want a different loop starvation, all you need to do is to adjust the `intensity` for  a couple of times, find out your sweet spot, and fix it there.

Hope this helps!

**Comment by logidelic** [comment]
Thanks for the response! I will use `intensity` for now as you suggest.

I still think it would be useful to allow specifying a threshold in ms as an alternative (since different server speeds will result in different timings with the same intensity). If I can find the time I may work on it.

Thank again!

**Comment by gireeshpunathil** [comment]
> I still think it would be useful to allow specifying a threshold in ms as an alternative

@logidelic - I understand, and have full desire to achieve that, as that is the best abstraction that an end user can understand, as opposed to the number of JSON fields being processed as a criteria. Unfortunately, there is no direct way of detecting elapsed time while performing CPU bound operations without resorting to heavy duty checks. Normal node.js timer primitives such as `setInterval` etc. will not work here, as those are async. and will not go well with the JSON processing.

In summary, desire is to expose a time threshold, but we are not yet there!

**Comment by logidelic** [comment]
Hmm, in my own code I just take the difference between two `process.hrtime.bigint()` to get timing... Are you suggesting that's too expensive?

**Comment by gireeshpunathil** [comment]
@logidelic - I haven't tested to see how expensive it is; but assuming it is not, still how frequently will you call it, to figure out an interval has reached? after every JSON field processing? or every 100 fields? there is a tradeoff between number of fields we want to skip versus the precision in meeting the responsiveness SLA versus the perf penalty we pay in calling the `hrtime`.

**Comment by aral** [comment]
> The value of `5 ms` is obtained from a normal linux system when we used `512` as the intensity. So we used this default value.

Hi @gireeshpunathil,

The `512` you mention here… is that different to the `intensity` argument we pass in the API? The readme states that the `intensity` parameter expects a value between 1 and 32. Are the docs incorrect or is this a different/internal value for intensity that you normalise to the public value for 1 (the default).

Thanks in advance :)

**Comment by gireeshpunathil** [comment]
@aral - thanks for asking.

https://github.com/ibmruntimes/yieldable-json/blob/fc2deec4b6430a839cd220dea3217cefa1728487/yieldable-parser.js#L167

as you can see, the `intensity` is the graded strength. Its product with 512 gives the actual number of iterations before we yield the loop.

So in summary, while the `intensity` gives a true measure of the intensity of work you do in an iteration, it is not 512.

**Comment by aral** [comment]
Thank you for the clarification, @gireeshpunathil :) Lovely module, by the way.
